apbs/use-shared-services-across-namespaces.md

```diff
@@ -0,0 +1,70 @@
+
+## Goals
+
+- A convenient way for the developer to tell the platform about an existing deployment of the service he is about to deploy.
+- The APB for a shared service automatically connecting to the shared service in another namespace, using the provided details.
+- The APB for a shared service automatically configuring the shared service in another namespace and exposing the configuration details to the current namespace.
+
+## Non-Goals
+
+- The platform intelligently discovering and connecting to services that are running in seperate namespaces.
+- The platform intelligently discovering and suggesting services to connect to.
+
+## Proposed Solution
+
+### Provision
+For services that are suited to being shared services (such as KeyCloak), we can make a change the apb.yml, allowing the developer to specify connection details for an already existing instance of the service this APB would otherwise deploy.
+
+Example:
+![Image of KeyCloak APB config screen with shared service fields](./images/shared-services-config.png)
+
+In this example, if the `Host of existing KeyCloak service` were left empty then a new service would be installed. However if this field were populated the APB would connect to the host with the provided auth details and then complete the APB as usual using this existing shared Service.
+
+Once the shared service has been configured via the APB, the APB will then expose these configuration details to the current namespace in it's usual manner (i.e. config-map, secret, etc).
+
+If the APB is unable to connect to the supplied Shared Service details then the APB should exit with a failure status and report the problem in a way that will make trouble-shooting as convenient as possible for the developer.
+
+Most, if not all, of this work will be executed inside the APB logic for services deemed suitable for conversion to a Shared Service. Possible candidates for this work are listed below, but this list is not final:
+- KeyCloak
+- UPS
+- Build Farm
+
+### Deprovision
+When running a deprovision of Connected Service, the APB should clear up it's own configuration from the shared service, but should not deprovision the shared service.
+
+For example, in namespace-a the KeyCloak APB would have connected to KeyCloak in namespace-b and created a realm, a client, and a bearer-client. This realm and users should be removed from KeyCloak in namespace-b and the secret they are stored in, in namespace-a. Keycloak in namespace-b will still be running when this is completed.
+
+When deprovisioning a Shared Service the service should be removed whether or not there are still existing Connected Services.
```

Comment:
I think we need to make this a little bit clearer, I know what you mean but could be hard to parse.
So essentially if the resources created during the provision are present in the namespace then it will deprovision them otherwise it will only remove service specific items (such as the clients etc) and not the actual running pods